src/tacacs/pam/0008-Extract-tacacs-support-functions-into-library.patch

```diff
@@ -0,0 +1,659 @@
+             , __FUNCTION__, PAM_TAC_VMAJ, PAM_TAC_VMIN, PAM_TAC_VPAT);
+ 
+     if ((user = _pam_get_user(pamh)) == NULL) {
+-        return PAM_USER_UNKNOWN;
++        return PAM_USER_UNKNOWN;
+     }
+ 
+     if (ctrl & PAM_TAC_DEBUG)
+@@ -589,7 +586,7 @@ int pam_sm_acct_mgmt (pam_handle_t * pamh, int flags,
+     if(tac_fd < 0) {
+         _pam_log (LOG_ERR, "TACACS+ server unavailable");
+         if(arep.msg != NULL) {
+-            free (arep.msg);
++            free (arep.msg);
+         }
+ 
+         return PAM_AUTH_ERR;
+@@ -732,9 +729,9 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+     syslog(LOG_DEBUG, "%s(flags=%d, argc=%d)", __func__, flags, argc);
+ 
+     if (   (pam_get_item(pamh, PAM_OLDAUTHTOK, &pam_pass) == PAM_SUCCESS)
+-        && (pam_pass != NULL) ) {
++        && (pam_pass != NULL) ) {
+          if ((pass = strdup(pam_pass)) == NULL) {
+-              return PAM_BUF_ERR;
++              return PAM_BUF_ERR;
+          }
+     } else {
+         pass = strdup("");
+@@ -743,7 +740,7 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+     if ((user = _pam_get_user(pamh)) == NULL) {
+         if(pass) {
+                 free(pass);
+-        }
++        }
+ 
+         return PAM_USER_UNKNOWN;
+     }
+@@ -828,7 +825,7 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+                     status = PAM_SUCCESS;
+                     communicating = 0;
+ 
+-                    active_server.addr = tac_srv[srv_i].addr;
++                    active_server.addr = tac_srv[srv_i].addr;
+                     /* copy secret to key */
+                     snprintf(active_server.key, sizeof(active_server.key), "%s", tac_srv[srv_i].key);
+ 
+diff --git a/pam_tacplus.h b/pam_tacplus.h
+index bc71b54..e7b30f7 100644
+--- a/pam_tacplus.h
++++ b/pam_tacplus.h
+@@ -31,12 +31,6 @@
+ #include <security/pam_appl.h>
+ #include <security/pam_modules.h>
+ 
+-/* pam_tacplus command line options */
+-#define PAM_TAC_DEBUG 0x01
+-#define PAM_TAC_ACCT  0x02 /* account on all specified servers */
+-#define PAM_TAC_USE_FIRST_PASS 0x04
+-#define PAM_TAC_TRY_FIRST_PASS 0x08
+-
+ /* pam_tacplus major, minor and patchlevel version numbers */
+ #define PAM_TAC_VMAJ 1
+ #define PAM_TAC_VMIN 3
+diff --git a/support.c b/support.c
+index 3d38f24..6b2e72d 100644
+--- a/support.c
++++ b/support.c
+@@ -29,6 +29,7 @@
+ 
+ #include <stdlib.h>
+ #include <string.h>
++#include <ctype.h> /* isspace() */
+ 
+ /* tacacs server information */
+ tacplus_server_t tac_srv[TAC_PLUS_MAXSERVERS];
+@@ -42,7 +43,7 @@ char tac_service[64];
+ char tac_protocol[64];
+ char tac_prompt[64];
+ char *__vrfname=NULL;
+-char tac_source_ip[64];
++char tac_source_ip[64];
+ 
+ /* source address */
+ struct addrinfo tac_source_addr;
+@@ -181,12 +182,33 @@ int tacacs_get_password (pam_handle_t * pamh, int flags
+ 
+     return PAM_SUCCESS;
+ }
+-
++
++/*
++ * Set tacacs server addrinfo.
++ */
++void set_tacacs_server_addr(int tac_srv_no, struct addrinfo* server) {
```

Comment:
>set_tacacs_server_addr [](http://example.com/codeflow?start=6&length=22)

Is this function your new code? If yes, is it possible not to use global variable?